torch/quantization/fx/_equalize.py

```diff
@@ -350,33 +376,84 @@ def scale_weight_node(
     scaled_bias = torch.mul(bias, next_equalization_scale)
     modules[node.target].bias = nn.Parameter(scaled_bias)
 
+def scale_weight_functional(
+    op_node: Node,
+    model: GraphModule,
+    modules: Dict[str, nn.Module],
+    equalization_scale: torch.Tensor,
+) -> None:
+    """ Scales the weight value for functional layers
+    """
+
+    # Find the next functional node so that we can construct the weight observer nodes
+    assert(isinstance(op_node.args[1], Node))
+    weight_observer_nodes = collect_producer_nodes(op_node.args[1])
+    if weight_observer_nodes is None:
+        return
+
+    for weight_node in weight_observer_nodes:
+        # Find the node containing the weight values
+        if weight_node.op == 'get_attr':
+            weight = model.get_buffer(str(weight_node.target))
+
+            # Scale the weights for input-weight equalization
+            # If the following layer needs to be equalized then we will multiply its scale
+            scaled_weight = torch.mul(weight, torch.reciprocal(equalization_scale))
+            # TODO: connecting functional layers?
+            # scaled_weight = torch.mul(scaled_weight, next_equalization_scale)
+
+            parent_name, name = _parent_name(weight_node.target)
+            setattr(modules[parent_name], name, scaled_weight)
+            assert(torch.allclose(model.get_buffer(str(weight_node.target)), scaled_weight))
+
+def clear_weight_quant_obs_node(op_node: Node, modules: Dict[str, nn.Module]) -> None:
+    """ Given the operation node, we want find the corresponding quantization
+    observer and reset its min/max values
+    """
+    assert(op_node.op == 'call_function')
+    assert(isinstance(op_node.args[1], Node))
+    weight_observer_nodes = collect_producer_nodes(op_node.args[1])
```

Comment:
I'm wondering if we should use more strict rules to trace back the weight observer nodes here, say we have:

weight_eqs_observer - weight_quant_observer - weight

should we just trace back two steps here?

Reply:
I think throughout the code we will eventually need to get every single node the `weight_observer_nodes` list. Could we find the `weight_observer_nodes` list once, pass it into functions where we need this observer node list, and then index into it based on what values we need? 

For example, we know the list should look something like: `[weight_eq_obs, weight_quant_obs, weight]`. So if we wanted `weight_quant_obs` and the value at `weight_observer_nodes[1]` was not the quantization observer, then we would return an error? We could maybe make a dictionary keeping track of the indices similar to [this](https://github.com/pytorch/pytorch/blob/master/torch/quantization/fx/utils.py#L17)?